### DISCUSSION TITLE
Database connection pool exhaustion error

### DISCUSSION THREAD

**Discussion by wbew** [SeaQL/sea-orm#893] [Q&A] [answered]
Hello! We're currently encountering the following error while performing several database queries in parallel:
```
An error occurred while performing a query: "Failed to acquire connection from pool.".
```

## Context
On `sea-orm v0.8.0`

We are using a Postgres database. We have `ConnectOptions` configured roughly as follows:
```rust
    static GENERAL_TIMEOUT_SECS: u64 = 10;

    opt.max_connections(10)
        .min_connections(5)
        .connect_timeout(Duration::from_secs(GENERAL_TIMEOUT_SECS))
        .idle_timeout(Duration::from_secs(GENERAL_TIMEOUT_SECS))
        .max_lifetime(Duration::from_secs(GENERAL_TIMEOUT_SECS))
        .sqlx_logging(false); 
```

While we can't share the exact details of the tables involved in this particular error, they're roughly as follows:

```
                           Table "public.table_foo"
        Column        |            Type             |               Modifiers               
----------------------+-----------------------------+---------------------------------------
 id                   | uuid                        | not null default uuid_function()
 name                 | character varying(255)      | not null
 key_uuid             | uuid                        | 
 another_id           | bigint                      | not null
 created_at           | timestamp without time zone | not null default now()
 created_by           | integer                     | not null
 modified_at          | timestamp without time zone | not null default now()
 modified_by          | integer                     | not null
 another_another_id   | integer                     | 
 another_key_uuid     | uuid 
```

```
                      Table "public.table_bar"
       Column       |            Type             |               Modifiers               
--------------------+-----------------------------+---------------------------------------
 uuid               | uuid                        | not null default uuid_function()
 object_uuid        | uuid                        | not null
 data_1             | character varying(255)      | not null
 data_2             | character varying(255)      | not null
 data_3             | text                        | not null
 created_at         | timestamp without time zone | not null default now()
 created_by         | integer                     | not null
```

```
                                            Table "public.table_baz"
         Column          |            Type             |                       Modifiers                        
-------------------------+-----------------------------+--------------------------------------------------------
 id                      | bigint                      | not null default nextval(...)
 uuid                    | uuid                        | not null
 category                | custom                      | default ...
 data_1                  | text                        | not null
 another_id              | bigint                      | not null
 is_active               | boolean                     | not null
 created_by              | bigint                      | 
 created_at              | timestamp without time zone | not null
 disabled_by             | bigint                      | 
 modified_at             | timestamp without time zone | not null
 modified_by             | bigint                      | 
 disabled_at             | timestamp without time zone | 
 data_2                  | text                        | 
 is_paused               | boolean                     | not null
 paused_by               | bigint                      | 
 paused_at               | timestamp without time zone | 
 resumed_by              | bigint                      | 
 resumed_at              | timestamp without time zone | 
 data_3                  | text                        | not null
 expires_at              | timestamp without time zone | 
```

And one more table very similar to `table_baz`.

## Hypothesis

We're running a GraphQL API server with `async_graphql` as the core library. In general, a GraphQL request can kick off many queries in parallel depending on how it resolves. 

In theory, `SeaORM`s connection pool abstraction should handle many parallel queries up to the connection limit by waiting for a free connection (up to 10 seconds as we've configured). When we encounter this error however, we see the entire GraphQL request occurs in a span of only a few hundred milliseconds. Interestingly, there are 11 (10 is our max) query executions in our trace monitoring. Some/all of them might be attempting to execute in parallel hence the failure. 

In other words, we think there is some edge case or non-graceful handling occurring such that `SeaORM` does not properly wait for connections to free up or does not respect the max connections. Before opening an official bug request, we were wondering if this sounds like it could be possible or if we've misconfigured something? Alternatively, could this be an issue in `sqlx`? Thanks!

**Comment by billy1624** [comment]
Hey @001wwang, thank you for the detail explanation!!

After some digging and investigation, the error was thrown when it fail to acquire connection from the connection pool:

https://github.com/SeaQL/sea-orm/blob/e85a859266dc6e8863e4bbb7a546fe36d88606cd/src/driver/sqlx_mysql.rs#L84-L95

By default SQLx connection pool will wait 30 seconds before it throw errors. But obviously 30 seconds is not enough for the edge case like above. Where large number of query execute concurrently.

The acquire timeout can be configured on [`sqlx::PoolOptions::acquire_timeout`](https://docs.rs/sqlx/latest/sqlx/pool/struct.PoolOptions.html#method.acquire_timeout) but this method isn't exposed on SeaORM. Which I think we should add that to SeaORM, I'll create a PR for that.

I have also added test cases to reproduce the above errors by inserting 10,000 rows concurrently. Then, fix the error by increasing acquire timeout to 90 seconds.

**Comment by billy1624** [comment, marked as answer]
I'll convert this into an issue

**Reply by billy1624** [reply]
https://github.com/SeaQL/sea-orm/issues/896

**Reply by wbew** [reply]
Thank you! I will close this out.
